### PR TITLE
Fix bug in empty!, add tests for WeakKeyDict

### PR DESCRIPTION
**Pull request by kshyatt**

### CODE REVIEW

File: base/weakkeydict.jl [hunk 1]
```diff
@@ -45,7 +45,7 @@ get!{K}(default::Callable, wkh::WeakKeyDict{K}, key) = lock(() -> get!(default,
 pop!{K}(wkh::WeakKeyDict{K}, key) = lock(() -> pop!(wkh.ht, key), wkh)
 pop!{K}(wkh::WeakKeyDict{K}, key, default) = lock(() -> pop!(wkh.ht, key, default), wkh)
 delete!{K}(wkh::WeakKeyDict{K}, key) = lock(() -> delete!(wkh.ht, key), wkh)
-empty!(wkh::WeakKeyDict) = (lock(() -> empty!(wkh.ht)); wkh)
+empty!(wkh::WeakKeyDict) = lock(() -> empty!(wkh.ht), wkh)
```
Comment (StefanKarpinski) [inline base/weakkeydict.jl]: Subtle. What was the bug here?

Comment (kshyatt) [inline base/weakkeydict.jl]:
```julia
  Got an exception of type MethodError outside of a @test
  MethodError: no method matching lock(::Base.##384#385{WeakKeyDict{Any,Any}})
  Closest candidates are:
    lock(::Any, !Matched::WeakKeyDict{K,V}) at weakkeydict.jl:21
    lock(::Any, !Matched::Any) at lock.jl:99
    lock(!Matched::Base.Threads.Mutex) at locks.jl:206
    ...
   in empty!; at ./weakkeydict.jl:48 [inlined]
```

Comment (vtjnash) [inline base/weakkeydict.jl]: Won't this return wkd.ht now, instead of wkd?

Comment (kshyatt) [inline base/weakkeydict.jl]: I can add a test for that? We can see.

Comment (kshyatt) [inline base/weakkeydict.jl]: Yes, it's returning `wkd.ht`. Is there an easy way to fix that?

Comment (Sacha0) [inline base/weakkeydict.jl]: Perhaps `empty!(wkh::WeakKeyDict) = (lock(() -> empty!(wkh.ht), wkh); wkh)`? Best!

Comment (kshyatt) [inline base/weakkeydict.jl]: That got it. Thanks, @Sacha0

**Comment by kshyatt** [comment]
CI (except for OSX) passed - OK to merge?

**Pull request merged**